File: clients/Cpp/MyStrategy.cpp
#include "MyStrategy.hpp"
#include <exception>

MyStrategy::MyStrategy() {}

Action MyStrategy::getAction(const PlayerView& playerView, DebugInterface* debugInterface)
{
    Action result = Action(std::unordered_map<int, EntityAction>());
    int myId = playerView.myId;
    for (size_t i = 0; i < playerView.entities.size(); i++) {
        const Entity& entity = playerView.entities[i];
        if (entity.playerId == nullptr || *entity.playerId != myId) {
            continue;
        }
        const EntityProperties& properties = playerView.entityProperties.at(entity.entityType);

        std::shared_ptr<MoveAction> moveAction = nullptr;
        std::shared_ptr<BuildAction> buildAction = nullptr;
        if (properties.canMove) {
            moveAction = std::shared_ptr<MoveAction>(new MoveAction(
                Vec2Int(playerView.mapSize - 1, playerView.mapSize - 1),
                true,
                true));
        } else if (properties.build != nullptr) {
            EntityType entityType = properties.build->options[0];
            size_t currentUnits = 0;
            for (size_t j = 0; j < playerView.entities.size(); j++) {
                if (playerView.entities[j].playerId != nullptr && *playerView.entities[j].playerId == myId
                    && playerView.entities[j].entityType == entityType) {
                    currentUnits++;
                }
            }
            if ((currentUnits + 1) * playerView.entityProperties.at(entityType).populationUse <= properties.populationProvide) {
                buildAction = std::shared_ptr<BuildAction>(new BuildAction(
                    entityType,
                    Vec2Int(entity.position.x + properties.size, entity.position.y + properties.size - 1)));
            }
        }
        std::vector<EntityType> validAutoAttackTargets;
        if (entity.entityType == BUILDER_UNIT) {
            validAutoAttackTargets.push_back(RESOURCE);
        }
        result.entityActions[entity.id] = EntityAction(
            moveAction,
            buildAction,
            std::shared_ptr<AttackAction>(new AttackAction(
                nullptr, std::shared_ptr<AutoAttack>(new AutoAttack(properties.sightRange, validAutoAttackTargets)))),
            nullptr);
    }
    return result;
}

void MyStrategy::debugUpdate(const PlayerView& playerView, DebugInterface& debugInterface)
{
    debugInterface.send(DebugCommand::Clear());
    debugInterface.getState();
}

File: clients/CSharp/MyStrategy.cs
using Aicup2020.Model;

namespace Aicup2020
{
    public class MyStrategy
    {
        public Action GetAction(PlayerView playerView, DebugInterface debugInterface)
        {
            Action result = new Action(new System.Collections.Generic.Dictionary<int, Model.EntityAction>());
            int myId = playerView.MyId;
            foreach (var entity in playerView.Entities) {
                if (entity.PlayerId != myId) {
                    continue;
                }
                EntityProperties properties = playerView.EntityProperties[entity.EntityType];
                MoveAction? moveAction = null;
                BuildAction? buildAction = null;
                if (properties.CanMove) {
                    moveAction = new MoveAction(
                        new Vec2Int(playerView.MapSize - 1, playerView.MapSize - 1),
                        true,
                        true
                    );
                } else if (properties.Build != null) {
                    EntityType entityType = properties.Build.Value.Options[0];
                    int currentUnits = 0;
                    foreach (var otherEntity in playerView.Entities) {
                        if (otherEntity.PlayerId != null && otherEntity.PlayerId == myId
                            && otherEntity.EntityType == entityType) {
                            currentUnits++;
                        }
                    }
                    if ((currentUnits + 1) * playerView.EntityProperties[entityType].PopulationUse <= properties.PopulationProvide) {
                        buildAction = new BuildAction(
                            entityType,
                            new Vec2Int(entity.Position.X + properties.Size, entity.Position.Y + properties.Size - 1)
                        );
                    }
                }
                EntityType[] validAutoAttackTargets;
                if (entity.EntityType == EntityType.BuilderUnit) {
                    validAutoAttackTargets = new EntityType[] { EntityType.Resource };
                } else {
                    validAutoAttackTargets = new EntityType[0];
                }
                result.EntityActions[entity.Id] = new EntityAction(
                    moveAction,
                    buildAction,
                    new AttackAction(
                        null,
                        new AutoAttack(properties.SightRange, validAutoAttackTargets)
                    ),
                    null
                );
            }
            return result;
        }

        public void DebugUpdate(PlayerView playerView, DebugInterface debugInterface) 
        {
            debugInterface.Send(new DebugCommand.Clear());
            debugInterface.GetState();
        }
    }
}